### ISSUE TITLE
Browserbox 2.*

### ISSUE THREAD

**Issue by felixhammerl** [emailjs/emailjs-imap-client#72]
Hi everyone,

here's a couple of things I wanna propose: ES6/2015 is here to stay and I'd suggest we embrace it. It offers some significant advantages over the old ways of doing things, specifically talking about generators and promises, as well as stuff as arrow functions, ...

**Meta stuff**
- [x] Move/Fork browserbox to the https://github.com/emailjs organization (postponed to 01/2016)
- [x] Rename it  (postponed to 01/2016)

**Code**
- [x] Remove optional callback API in favor of Promises
- [x] Refactor browserbox.js, use Promises internally
- [x] Refactor browserbox-imap.js, use Promises internally (where useful)
- [x] Remove `next()`
- [x] Explore usage of Streams in Browserbox
- [x] Remove axe
- [x] Pass logging function as option, defaults to console.log/error
- [x] Remove browser globals
- [x] Throw out phantomjs
- [x] Return from `connect()` when client is up and running, i.e. when `onauth()` is done, remove `onauth()` callback from api
- [x] Return from `close()` when connection is closed, i.e. when `onclose()` is done
- [x] Remove `onclose()` callback from api: connection unexpectedly is an error and should be handled as such.
- [x] Update documentation
- [ ] Enhance error reporting with more detailed failure codes
- [x] Avoid dissociation of errors in _processServerQueue, see #25

**Comment by tanx** [comment]
😁👍🏼

**Comment by asutherland** [comment]
If undertaking a refactor/restructuring of control flow and APIs, I'd suggest adopting the use of WHATWG streams (https://github.com/whatwg/streams) throughout the codebase.  Although not integrated in master yet, the Firefox OS Gaia email client has a branch where we've adopted streams to help improve our memory usage when downloading large Blobs.

For IMAP and browserbox, the implementation uses the "pull" mechanism supported by streams to do chunked downloads, but it would be fantastic if browserbox FETCH requests could return a stream of (message) objects.  Literals corresponding to body-parts could also be streams provided as part of the object, which could then be piped-through a mime parser to decode the stream as it arrives, etc.

As an example of the core of @mcav's work on this, please see https://github.com/mcav/gaia-email-libs-and-more/blob/streaming/js/mime-streams.js and perhaps other things in the same commit like the refactoring of our POP3 implementation to use streams.  (Note that those changes use jsmime for mime parsing rather than the mimeparser lib.  Also note that the code is now being refactored slightly as part of its integration into our conversations-centric branch, which includes niceties like splitting the larger file out into smaller per-class/method, etc.)

Aside: The existing automagic promise generation when a callback is not provided works pretty well for consumers.

**Comment by felixhammerl** [comment]
For the Promises thingy, I'm not talking about the external API, which indeed works quite ok and would probably not change too much. I'd wanna throw out the optional callback API.

Maybe that's just me, but if we re-work this codebase, I'd wanna go for something that is actually implemented in the browsers. Streams is too brittle for my taste, but it does look good and it it comes to be supported, I'm all for adding it later. This way, we don't have to polyfill and/or transpile the code.

I'd also suggest throwing out UMD in favor of CommonJS for the time being. Most projects have a build pipeline with some sort of package build tooling anyway, whether this is webpack or browserify or something similar...

**Comment by asutherland** [comment]
#### Streams ####

My impression of the streams spec is that it has largely stabilized.  Chrome has even implemented readable stream support in its fetch implementation (https://groups.google.com/a/chromium.org/forum/#!topic/blink-dev/35_QSL1ABTY).  And unlike Promises which required special browser engine support for semantics with the micro-task queues and all that, streams don't require browser support for them to be useful.  Browser support will potentially bring a lot of efficiency wins in terms of reduced memory usage, explicit back-pressure support, and data generally touching less threads, of course.

Having overhauled our codebase for conversations to use generators and now slightly for streams, I've found they both favor different control-flow idioms, so it seems beneficial to just jump directly to streams.  Also, if you use generators/etc. and you're avoiding transpiling, you're going to need a generator driver like https://github.com/tj/co; it's possible the stream model may eliminate that need.

While my impression is that whiteout has never particularly targeted constrained-resource devices, even a beefy device could likely benefit from efficient streaming if downloading multiple messages in parallel that have 20+ megs of attachments on them with aggressive attachment downloading, etc.  My understanding is that the email.js stack right now would potentially produce a *lot* of garbage as the 20 meg literal makes each way through each stage of the processing pipeline in that single large chunk.

#### AMD/UMD ####

As a contributor to browserbox who uses AMD-form in our trees (and without a build step during many dev scenarios), it's been very handy for me to be able to directly copy browserbox*.js to/from my tree or symlink it while during debugging and testing.  (Specifically, it lets me treat only one location as the current "truth" and avoid having to perform manual merges/patching between repositories.  It's great for letting me iterate by implementing mocha tests in browserbox that pass first, then immediately testing in a real-world scenario and iterating quickly to add more mocha tests, etc.)

I do see that the current UMD used by browserbox is a bit of a hassle since it requires the dependencies to be specified 3 times, and I do think I've seen commits in the commit log that indicate that this has been a source of bugs.  How would you feel about using one of the variants that allows inline require calls and thereby only one set of require()s?  The browser globals case loses out there, but arguably if there is a goal to support the browser global case, webpack/browserify can be used to create a single UMD build that expose a single BrowserBox global.  (Or does its shim when used as a global expose everything?)

For example, our AMD code in our codebase now looks like the following most of the time:
```js
define(function(require, exports) {
var modA = require('modA');
var modB = require('modB');
});
```

The AMD loader does need to regexp out the deps if there's no build step, but it does works at dev-time without a build step this way.  The idioms for this are https://github.com/umdjs/umd/blob/master/templates/commonjsAdapter.js and https://github.com/umdjs/umd/blob/master/templates/nodeAdapter.js and there may be others too.

**Comment by asutherland** [comment]
Er, and actually, I should say, we now try to do:
```js
define(function(require) {
var modA = require('modA');
return {
  exportedThingOne,
  exportedThingTwo
};
});
```

But explicitly requiring exports (and possibly module) works too.

**Comment by tanx** [comment]
If I may make a suggestion about the module syntax discussion. It's possible to use only common.js syntax for browserbox and build a UMD module with browserify. We do this for OpenPGP.js and simply commit the built module as a single script file to the '/dist' directory in the git repo for each release. That way users of the library aren't burdened with a build step and the codebase can use only common.js syntax at development time.

**Comment by asutherland** [comment]
Yes, a lot of CJS-styled modules I consume in various AMD projects I have do that too, and frequently it works great.  I think it can work less well when there are shared dependencies that need not be multiply loaded (ex: https://github.com/tj/co if you're using that to drive generators), or that need to be clobbered (ex: we replace the "axe" modules in our namespace with shims to our logging layer.)  OTOH tj/co isn't huge, and if https://github.com/whiteout-io/browserbox/issues/71 is resolved by letting us pass in our own logger function, that's probably fine too.

My main point was just that having no build step has been really useful for me as a contributor.  It's not the end of the world if there's a build/wrap step, it just introduces some extra effort with related activation energy/build hazards that aren't there right now.  I and the rest of the FxOS mail team are probably the only people using AMD and browserbox who would care, and we're still going to try and upstream fixes no matter what the upstream module system is, so please take what I say with the appropriate level of salt.  (Hope that's not too idiomatic; not sure of another appropriate phrasing. :smile: :)

**Comment by tanx** [comment]
> My main point was just that having no build step has been really useful for me as a contributor.

Ok gotcha.

**Comment by felixhammerl** [comment]
**UMD**: As @asutherland already mentioned, there are a couple of issues involved when using UMD w/ stuff like electron or nwjs (née Node-Webkit). I totally get that being able to just drop browserbox in without a build step is beneficial. Let's tackle that if and when that becomes a problem.

**Logging**: I definitely want to purge axe from the face of the earth, possibly by being able to pass a logging function that adheres to a specific interface. No idea why I didn't think of that in the beginng ... 

**Streams**: 

> My understanding is that the email.js stack right now would potentially produce a lot of garbage as the 20 meg literal makes each way through each stage of the processing pipeline in that single large chunk.

Yep ... \*shame\* ... I'm gonna take some time to rethink the current architecture and how that could be achieved with generators and streams. if generators, i planned on using tj/co as a generator driver. there's a **lot** of garbage collection involved in the current stack, which i wanted to eliminate one by one, streams may help with that, not so sure for generators. back-pressure support is definitely beneficial... 

We're not gonna tackle all of the issues at once, so we might a

**Comment by andris9** [comment]
Regarding meta stuff, I'd suggest forking, not moving, related modules to the emailjs account as I don't think we have the rights to move stuff away from Whiteout account. Renaming is ok and should be done anyway to clean things up since some of the modules have wo-* prefix in npm, additionally gihtub names do not match with npm names etc. Maybe prefix everything with emailjs-* like *emailjs-addressparser*, *emailjs-browserbox*?

I agree that axe should be replaced with a more general logger. I have usually used [bunyan](https://github.com/trentm/node-bunyan) compatible logger objects, where I use a defined subset of bunyan, so I can either pass an actual bunyan logger object or a custom small logger implementation, see [smtp-server](https://github.com/andris9/smtp-server/blob/6ccb74956603f01e8480b77a2f2b4e032fb74578/lib/smtp-server.js#L196-L227) for example. In this case the logger interface looks like this:

    logger [info | debug | error] ( log_message, ...sprintf_args )

**Comment by tanx** [comment]
@andris forking would have the disadvantage of loosing all git repo meta resources like issues. I would be in favor of moving from whiteoutio to emailjs as the components are branded as part of that effort anyway.

Regarding the future of the whiteoutio github organization. There will be a public announcement shortly and all MIT licensed components will stay available one way or another.

**Comment by felixhammerl** [comment]
hi guys,

i started the refactorung on branch https://github.com/whiteout-io/browserbox/tree/dev/v2. so far, i ported the external api to promises, which turned out to be pretty painless. all in all, i *really* like Promises and how the refactoring is going so far, just check this stuff out: [_onReady before refactoring](https://github.com/whiteout-io/browserbox/blob/master/src/browserbox.js#L204-L236) versus [_onReady after refactoring](https://github.com/whiteout-io/browserbox/blob/dev/v2/src/browserbox.js#L209-L222).
I also was able to remove the infamous `next()` callback that was omnipresent in the code. and, oh, the main js file is 330 lines shorter now ;)
that being said, it is still work in progress. a lot left to do in there :)


the only thing that i can't seem to figure out is why one of the commands using the `precheck` api does not work. the latest commit on dev/v2 breaks the build for this reason. 


@andris9: do you understand what's going wrong in the integration test here? seems like the command is not run on the correct folder, but for all the other commands, the test seems to work just fine ... the relevant part of the integration test is here: https://github.com/whiteout-io/browserbox/blob/0ef70e60cd9db26c6674aad799f5d4476eccfe72/test/integration/browserbox-test.js#L461-L465

**Comment by felixhammerl** [comment]
Guys, do we still need the error codes thing in v2?

**Comment by felixhammerl** [comment]
You might want to check this out: https://github.com/whiteout-io/browserbox/pull/75

**Comment by felixhammerl** [comment]
Oh, one more thing: this changes the version straight to v2.* ... i.e. skipping v1, which doesn't really make sense ... or we make this another minor 0.xx release, but then again, it completely breaks the API in a lot of places...

**Comment by andris9** [comment]
Skipping to v2.0.0 is fine. Using 0.x is not encouraged anymore anyway, so we shouldn't keep using 0.x versions.

**Comment by felixhammerl** [comment]
Concerning moving/forking, let's postpone that until January for organizational reasons. We can keep working on the libraries in this repository regardless...
